sway/tree/layout.c

```diff
@@ -619,30 +626,94 @@ static struct sway_container *get_swayc_in_direction_under(
 						wrap_candidate = parent->children->items[0];
 					}
 					if (config->force_focus_wrapping) {
-						return wrap_candidate;
+						 return sway_seat_get_focus_by_type(seat, wrap_candidate, C_VIEW);
 					}
 				}
 			} else {
 				wlr_log(L_DEBUG,
 					"cont %d-%p dir %i sibling %d: %p", idx,
 					container, dir, desired, parent->children->items[desired]);
-				return parent->children->items[desired];
+				return sway_seat_get_focus_by_type(seat,
+						parent->children->items[desired], C_VIEW);
 			}
 		}
 
 		if (!can_move) {
 			container = parent;
 			parent = parent->parent;
-			if (!parent || container == limit) {
+			if (!parent) {
 				// wrapping is the last chance
 				return wrap_candidate;
 			}
 		}
 	}
 }
 
-struct sway_container *container_get_in_direction(
-		struct sway_container *container, struct sway_seat *seat,
-		enum movement_direction dir) {
-	return get_swayc_in_direction_under(container, dir, seat, NULL);
+struct sway_container *container_replace_child(struct sway_container *child,
+		struct sway_container *new_child) {
+	struct sway_container *parent = child->parent;
+	if (parent == NULL) {
+		return NULL;
+	}
+	int i = index_child(child);
+
+	// TODO floating
+	parent->children->items[i] = new_child;
+	new_child->parent = parent;
+	child->parent = NULL;
+
+	// Set geometry for new child
+	new_child->x = child->x;
+	new_child->y = child->y;
+	new_child->width = child->width;
+	new_child->height = child->height;
+
+	// reset geometry for child
+	child->width = 0;
+	child->height = 0;
+
+	return parent;
+}
+
+struct sway_container *container_split(struct sway_container *child,
+		enum sway_container_layout layout) {
+	// TODO floating: cannot split a floating container
+	if (!sway_assert(child, "child cannot be null")) {
+		return NULL;
+	}
+	struct sway_container *cont = container_create(C_CONTAINER);
+
+	wlr_log(L_DEBUG, "creating container %p around %p", cont, child);
+
+	cont->prev_layout = L_NONE;
+	cont->layout = layout;
+	cont->width = child->width;
+	cont->height = child->height;
+	cont->x = child->x;
+	cont->y = child->y;
+
+	/* Container inherits all of workspaces children, layout and whatnot */
```

Comment:
Doesn't splitting a container just change the container layout? This is being handled at the command level iiuc

Reply:
Probably better to handle here than at the command level though tbh

Reply:
i3 is actually creating a container around the containers in the workspace in this case.

Reply:
hm